Orphaned DB entries for manually relocated audiobook files

### Discussed in https://github.com/orgs/grimmory-tools/discussions/1173

<div type='discussions-op-text'>

<sup>Originally posted by **jdeitsch** May  6, 2026</sup>
### What went wrong?

I added some audiobook files to my main eBook library and when I imported them in, they were not organized to my liking, so I manually moved the files to a new location and added that as a separate library.  When I went back to remove the entries for the files that were created during the import process, I could not remove them because the files were not found.

This was several days ago on v3.0.3, today I updated to the nightly build in hopes of a resolution being in place therein.  It was not.

Whenever I try to delete the entries, I either get a very messy error referencing every book ID entry and a type `0` error, or "An Unexpected Error Occurred" pop-up, and then I need to log back in because something goes weird with the session management (I assume this is due to the Java error, but I can also blame the teapot in space).

### How can we reproduce it?

Setup: Existing eBook-only library set up and organized

1. Add several audiobooks (both single-file and file-per-chapter) to the root directory of that library
2. Hit "re-scan"
3. Start organizing the newly imported files
  3a. Lose hope in your life [optional]
4. Manually relocate the audiobook files to another location for more compartmentalized colocation
5. Attempt to remove the now-orphaned listings for the books
6. Observe error


### What Build of Grimmory are you on?

Stable

### Your setup

- Grimmory Version: Occurred on 3.0.3 as well as Stable
- Docker Build
- Happens with Zen, Chrome, Firefox, Edge, and Arc browsers
- The Docker instance is running on Ubuntu Server, I am accessing from Windows 11

### Screenshots or error messages (optional)

Docker logs contain many variants of this error.  From what I can get from it, it cannot remove the entry because it cannot find the file, but my Java knowledge is roughly in line with my Sanscrit (~0.02% total knowledge)

```
Caused by: org.hibernate.StaleObjectStateException: Batch update returned unexpected row count from update 0 (expected row count 1 but was 0) [update book_file set added_on=?,archive_type=?,bitrate=?,book_id=?,book_type=?,channels=?,chapter_count=?,chapters_json=?,codec=?,current_hash=?,description=?,duration_seconds=?,file_name=?,file_size_kb=?,file_sub_path=?,is_folder_based=?,initial_hash=?,is_book=?,is_fixed_layout=?,sample_rate=? where id=?] for entity [org.booklore.model.entity.BookFileEntity with id '671']
        at org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard.staleObjectStateException(UpdateCoordinatorStandard.java:1065) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.persister.entity.mutation.UpdateCoordinatorStandard.lambda$doStaticUpdate$2(UpdateCoordinatorStandard.java:812) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.batch.internal.BatchImpl.checkRowCounts(BatchImpl.java:311) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$0(BatchImpl.java:268) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:63) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:247) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:226) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:194) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:658) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:505) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:381) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:138) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.internal.SessionImpl.fireFlush(SessionImpl.java:1453) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:484) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2080) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2002) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:426) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:166) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commitNoRollbackOnly(JdbcResourceLocalTransactionCoordinatorImpl.java:251) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:241) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:89) ~[hibernate-core-7.3.2.Final.jar!/:7.3.2.Final]
        at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:553) ~[spring-orm-7.0.7.jar!/:7.0.7]
        ... 162 common frames omitted
```
<img width="552" height="957" alt="image" src="https://github.com/user-attachments/assets/f9c34665-2d40-4736-a65f-8a931a2b37d2" />


### Before submitting

- [X] I've searched existing discussions and this hasn't been reported yet
- [X] I've reviewed this report and have completed it as fully as I can</div>

--- a/backend/src/main/java/org/booklore/service/book/BookService.java
+++ b/backend/src/main/java/org/booklore/service/book/BookService.java
@@ -1,5 +1,6 @@
 package org.booklore.service.book;
 
+import jakarta.persistence.EntityManager;
 import jakarta.servlet.http.HttpServletResponse;
 import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
@@ -52,6 +53,7 @@ public class BookService {
 
     private final BookRepository bookRepository;
     private final BookFileRepository bookFileRepository;
+    private final EntityManager entityManager;
     private final PdfViewerPreferencesRepository pdfViewerPreferencesRepository;
     private final CbxViewerPreferencesRepository cbxViewerPreferencesRepository;
     private final NewPdfViewerPreferencesRepository newPdfViewerPreferencesRepository;
@@ -461,6 +463,12 @@ public class BookService {
         }
 
         bookRepository.deleteAllInBatch(books);
+
+        // Because this is `InBatch` we need to clear and flush the entity manager to
+        // prevent unexpected updates of records when the transaction commits.
+        entityManager.clear();
+        entityManager.flush();
+
         auditService.log(AuditAction.BOOK_DELETED, "Deleted " + ids.size() + " book(s)");
         BookDeletionResponse response = new BookDeletionResponse(ids, failedFileDeletions);
         return failedFileDeletions.isEmpty()

--- a/backend/src/main/java/org/booklore/service/book/BookService.java
+++ b/backend/src/main/java/org/booklore/service/book/BookService.java
@@ -466,8 +466,8 @@ public class BookService {
 
         // Because this is `InBatch` we need to clear and flush the entity manager to
         // prevent unexpected updates of records when the transaction commits.
-        entityManager.clear();
         entityManager.flush();
+        entityManager.clear();
 
         auditService.log(AuditAction.BOOK_DELETED, "Deleted " + ids.size() + " book(s)");
         BookDeletionResponse response = new BookDeletionResponse(ids, failedFileDeletions);

--- a/backend/src/test/java/org/booklore/BookServiceDeleteTests.java
+++ b/backend/src/test/java/org/booklore/BookServiceDeleteTests.java
@@ -1,5 +1,6 @@
 package org.booklore;
 
+import jakarta.persistence.EntityManager;
 import org.booklore.config.security.service.AuthenticationService;
 import org.booklore.mapper.BookMapper;
 import org.booklore.repository.*;
@@ -57,10 +58,12 @@ class BookServiceDeleteTests {
         SidecarMetadataWriter sidecarMetadataWriter = Mockito.mock(SidecarMetadataWriter.class);
         FileStreamingService fileStreamingService = Mockito.mock(FileStreamingService.class);
         AuditService auditService = Mockito.mock(AuditService.class);
+        EntityManager entityManager = Mockito.mock(EntityManager.class);
 
         bookService = new BookService(
                 bookRepository,
                 bookFileRepository,
+                entityManager,
                 pdfViewerPreferencesRepository,
                 cbxViewerPreferencesRepository,
                 newPdfViewerPreferencesRepository,

--- a/backend/src/test/java/org/booklore/service/book/BookServiceTest.java
+++ b/backend/src/test/java/org/booklore/service/book/BookServiceTest.java
@@ -1,5 +1,6 @@
 package org.booklore.service.book;
 
+import jakarta.persistence.EntityManager;
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.servlet.http.HttpServletResponse;
 import org.booklore.config.security.service.AuthenticationService;
@@ -75,6 +76,8 @@ class BookServiceTest {
     private BookUpdateService bookUpdateService;
     @Mock
     private AuditService auditService;
+    @Mock
+    private EntityManager entityManager;
 
     @InjectMocks
     private BookService bookService;
@@ -359,6 +362,37 @@ class BookServiceTest {
             fileUtilsMock.when(() -> FileUtils.getBookFullPath(entity)).thenReturn(Path.of("/tmp/nonexistentfile.txt"));
             assertThrows(APIException.class, () -> bookService.getBookContent(12L));
         }
+    }
+
+    @Test
+    void deleteBooks_flushesEntityManager() throws Exception {
+        BookEntity entity = new BookEntity();
+        entity.setId(11L);
+        LibraryEntity library = new LibraryEntity();
+        library.setId(42L);
+        LibraryPathEntity libPath = new LibraryPathEntity();
+        libPath.setPath("/tmp");
+        library.setLibraryPaths(List.of(libPath));
+        entity.setLibrary(library);
+        entity.setLibraryPath(libPath);
+        BookFileEntity primaryFile = new BookFileEntity();
+        primaryFile.setBook(entity);
+        primaryFile.setFileSubPath("");
+        primaryFile.setFileName("bookfile.txt");
+        entity.setBookFiles(List.of(primaryFile));
+
+        Path filePath = Paths.get("/tmp/bookfile.txt");
+        Files.createDirectories(filePath.getParent());
+        Files.write(filePath, "abc".getBytes());
+
+        doNothing().when(bookRepository).deleteAllInBatch(anyList());
+        when(bookQueryService.findAllWithMetadataByIds(Set.of(11L))).thenReturn(List.of(entity));
+        when(authenticationService.getAuthenticatedUser()).thenReturn(testUser);
+
+        bookService.deleteBooks(Set.of(11L)).getBody();
+
+        verify(entityManager).flush();
+        verify(entityManager).clear();
     }
 
     @Test

--- a/backend/src/test/java/org/booklore/service/book/BookServiceTest.java
+++ b/backend/src/test/java/org/booklore/service/book/BookServiceTest.java
@@ -365,7 +365,7 @@ class BookServiceTest {
     }
 
     @Test
-    void deleteBooks_flushesEntityManager() throws Exception {
+    void deleteBooks_clearsEntityManager() throws Exception {
         BookEntity entity = new BookEntity();
         entity.setId(11L);
         LibraryEntity library = new LibraryEntity();
@@ -391,7 +391,6 @@ class BookServiceTest {
 
         bookService.deleteBooks(Set.of(11L)).getBody();
 
-        verify(entityManager).flush();
         verify(entityManager).clear();
     }
 

